src/sources.js

```diff
@@ -70,35 +74,80 @@ class Source {
   }
 }
 
-class Dataset {
+class Resource {
   constructor(source, pathParts) {
     this.source = source;
     this.pathParts = pathParts;
 
-    // Require baseParts, otherwise we have no actual dataset path.  This
-    // inspects baseParts because some of the pathParts (above) may not apply,
-    // which each Dataset subclass determines for itself.
+    // Require baseParts, otherwise we have no actual dataset/narrative path.
+    // This inspects baseParts because some of the pathParts (above) may not
+    // apply, which each Dataset/Narrative subclass determines for itself.
     if (!this.baseParts.length) {
-      throw new NoDatasetPathError();
+      throw new NoResourcePathError();
     }
   }
   get baseParts() {
     return this.pathParts.slice();
   }
-  baseNameFor(type) {
-    const baseName = this.baseParts.join("_");
-    return type === "main"
-      ? `${baseName}.json`
-      : `${baseName}_${type}.json`;
+  get baseName() {
+    return this.baseParts.join("_");
   }
-  async urlFor(type, method = 'GET') { // eslint-disable-line no-unused-vars
-    const url = new URL(this.baseNameFor(type), await this.source.baseUrl());
-    return url.toString();
+  async exists() {
+    throw new Error("exists() must be implemented by Resource subclasses");
+  }
+  subresource(type) { // eslint-disable-line no-unused-vars
+    throw new Error("subresource() must be implemented by Resource subclasses");
+  }
+}
+
+class Subresource {
+  constructor(resource, type) {
+    if (this.constructor === Subresource) {
+      throw new Error("Subresource interface class must be subclassed");
+    }
```

Comment:
Yep. It's checking that we aren't directly instantiating the Subresource class itself, since there are required methods it doesn't implement. Those would error if called, but possibly much later and deeper into a call stack than initial instantiation. it's a belt-and-suspenders, defensive approach.